Add private config to broadcast rank0 decision from the partitioner to all ranks

fmassa opened the pull request:
Summary: This PR adds a private configuration to the partitioner that ensures that the decision taken is the same across all ranks. This is a temporary workaround, as when size_hints are also taken into account in compiler collectives this workaround will not be needed anymore.

Test Plan:
This has been tested on some internal models, but I haven't added any tests in PyTorch (yet?)
T

Differential Revision: D73666017

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -1921,6 +1921,43 @@ def estimate_for_budget(b):
+        inputs = [x.name for x in joint_graph.find_nodes(op="placeholder")]
+        all_inputs = [None for _ in range(torch.distributed.get_world_size())]
+        with no_dispatch(), unset_fake_temporarily():
+            torch.distributed.all_gather_object(all_inputs, inputs)
```
can we either:

(1) tweak this collective to just compute a hash on the graph (without symint metas) and check that the hashes are the same

(2) if not, add a comment calling out the failure mode here and the potential followup?

bdhirsh reviewed (approved):
Otherwise this sounds ok to me

fmassa commented:
@pytorchbot merge

fmassa committed:
Add private config to broadcast rank0 decision from the partitioner to all ranks (#152264)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152264

This PR adds a private configuration to the partitioner that ensures that the decision taken is the same across all ranks. This is a temporary workaround, as when size_hints are also taken into account in compiler collectives this workaround will not be needed anymore.

Test Plan:
This has been tested on some internal models, but I haven't added any tests in PyTorch (yet?)
T

Reviewed By: bdhirsh

Differential Revision: D73666017

fmassa commented:
@pytorchbot merge
